### PR TITLE
Actor

### PR DESCRIPTION
**Pull request by wuminzhe**
#70

### CODE REVIEW

**Comment by clearloop** [comment]
This actor model implementation takes more code and more dependencies, and make the difficulty of developing this repo harder.

---

### 0. Avoid using `Arc` and `Mutex`

If you are trying to avoid using `Arc` and `Mutex`, please read the source code of `actix`. If you are trying to learn the actor model, please do it on your personal project, this project is under production.

### 1. Why `shadow` depends on `actix`?

Because it uses `actix-web` to serve APIs

### 2. Why `actix-web` depends on `actix`?

Because **it is a web framework**, and it contains a lot of thread interacting operation.

### 3. When to use big dependency like `actix`?

If you are writing a web framework, writing some services closing to the basic protocol, it is a good choice, for example, you can write a `Redis` in rust using `actix`.

**Comment by wuminzhe** [comment]
1. I never said I want to remove Arc
2. I want to remove the shared data, and improve the immediate processing capability.
3. I want to do things right. I believe, first, more deps and less codes, second, more mature libs and less self-made wheels. Maybe I'm wrong, and you don't need to accept my point.
4. I am not asking for this pr to be merged in, but for everyone to check the difference. If most of us think it's good, then merge, if bad, then give up. 
5. Making a standalone rep may be a good choice.

**Comment by clearloop** [comment]
> 5. Making a standalone repo may be a good choice.

I Accept this idea.

> 2. I want to remove the shared data, and improve the immediate processing capability.
> 3. I want to do things right. I believe, first, more deps and less codes, second, more mature libs and less self-made wheels. Maybe I'm wrong, and you don't need to accept my point.

But anyway, shared data is a basic operation in any system programming language, it's not a wheel, it spends few lines.

**Comment by wuminzhe** [comment]
Reducing the use of shared memory is one of the important means of decoupling. There is no good shared memory but only compromised shared memory. You designed the service abstraction layer, that's good, but very coupled. for example, all services depend on Config, all services depends on shared memory. The most basic problem is that it is difficult to test independently. When the system becomes complex, the benefits of decoupling will gradually become apparent.

And, modern programming facilities provide easier ways to solve the shared data race problem, such as databases, such as goroutines and actors, such as immutable variables. I just want to give it a try, is this not allowed?

The actor approach is this project is not necessarily correct, and there may be some problems that I have not encountered. You can ask specific questions about the code, for example, is it too inefficient to do this here, or what problems will it cause there. Also, you can compile this branch and try to run it to find the problems.

**Comment by clearloop** [comment]
> The actor approach is this project is not necessarily correct, and there may be some problems that I have not encountered. You can ask specific questions about the code, for example, is it too inefficient to do this here, or what problems will it cause there. Also, you can compile this branch and try to run it to find the problems.

I'm not baning you to try actor on this project, I don't have the right, but I know this is not the right way, for both bridger and actor, It's hard to tolerate that projects I have involved into turning into a wrong way, everything I'm saying or I have said is only for the code or project, I respect you very much sincerely, and I feel happy about someone has different ideas with me so I can think about if I am wrong.

Lots of ideas you have got are caused by you have few experiences in Rust, I have more than 2 years' experience writing Rust, and I had written almost 30 projects using rust, some of them are very big, framework level, so I know why you are thinking about these, and I know little things about the rust way, about which solution is closing to perfect.

I feel anxious because there are lots of issues that need to be solved in bridger and shadow which are more important than trying something, and you will master rust soon after these two projects, and you can do any trying efficiently at that time, finally.

> you can compile this branch and try to run it to find the problems.

I don't have enough time to do it because I'm out of the game now, I have some other things to do, but anyway, I will keep contributing to the projects I have involved in just for respecting the code I had written, of course, not always the \`banning\` thing.

> Reducing the use of shared memory is one of the important means of decoupling. 

You can chase decoupling if you are writing a framework like `substrate`, bridger is a little project.

> When the system becomes complex

Not all projects have to be complex, you are thinking about the complex way because you are not thinking about the light way.

> The most basic problem is that it is difficult to test independently

You can test every service by mocking the shared data, so will you do this when you change the arch, but mock the data which is not shared.

The difficulty of the testing is not caused by the arch, but the APIs.

> And, modern programming facilities provide easier ways to solve the shared data race problem, such as databases, such as goroutines and actors, such as immutable variables. I just want to give it a try, is this not allowed?

You will solve the data race problem easily when you understand the deep logic of shared memory, using outside dependencies only can make you pretend you are solving the problem, but the answer is no.

I will use a database if I'm writing projects like the backend of Subscan, goroutine is the go way, not rust, and actor is the rust way, but not for projects like the bridger.

It's like, we are trying to write a web page contains `<p>hello, world!</p>`,  you want to use `react/vue` to implement it because you want to solve the DOM update problem, the answer is, we just need to write a `<p>hello, world</p>`.

**Comment by wuminzhe** [comment]
You have a long history of rust development, I also do a development job for many years. But until today I dare not say which way to design or develop software is correct. 

And, your helloworld example is not appropriate to put here.

**Comment by clearloop** [comment]
> You have a long history of rust development, I also do a development job for many years. But until today I dare not say which way to design or develop software is correct.

I dare to say I'm approaching the perfect because I'm always trying to approach the perfect for the little 3 years of my developing history, every project I have written I spent most of the time thinking and trying to make them as perfect as I can imagine, as I can implement, I dare to say it because I always do my best.

**Comment by HackFisher** [comment]
@wuminzhe LGTM, but CI failed, do we need rust nightly for the CI.

@AurevoirXavier Would you please help review this PR, thanks.

**Comment by wi1dcard** [comment]
Can everyone cool down a bit please? After reading and trying to understand all the comments, I was still no idea about the advantages and disadvantages of this PR. I'm getting confused.

I assume that the current conflict is - should we use shared memory or not.

To push this PR (and any other projects) forward, people who support using shared memory or not, may leave their opinion of:

- Advantages of using shared memory (or another solution).
- Disadvantages of using shared memory (or another solution).
- Any documentation, paper, researching, reference that can support your opinion.

In summary, I believe in GitHub Pull Requests, developers coming here focus on collaborate instead of discuss the "right way or perfection of developing". Every individual developer can have his/her own logic and definition of "good" software, based on his/her experience of coding, including me. But, it would be nice if we can constantly sticking on the topic, and spend no more time to judge the design pattern without any clear, targeted, or reproducable issue.

**Comment by clearloop** [comment]
> I assume that the current conflict is - should we use shared memory or not.

Both solutions use shared memory, the difference is, the current solution in the master branch uses it directly, and this pr imports a dependency to use it.

> To push this PR (and any other projects) forward, people who support using shared memory or not, may leave their opinion of:
> 
> Advantages of using shared memory (or another solution).
> Disadvantages of using shared memory (or another solution).
> Any documentation, paper, researching, reference that can support your opinion.

Furthermore, the current solution is less than 10 lines in the importing dependency which has more than 10,000 lines indeed.

**Review by HackFisher — APPROVED** [review]

**Review by HackFisher — APPROVED** [review]

**Comment by wuminzhe** [comment]
There is no need to argue about these. I don't think this pr is so good, there are problems, but not you mentioned.
